### PR TITLE
[FEATURE] Centralized foundation model configuration for BYOKG-RAG

### PR DESCRIPTION
**Pull request by mykola-pereyma**
## Description

Closes #176

Extracts foundation model configuration into a centralized `ByoKGConfig` singleton for BYOKG-RAG, following the same pattern as `GraphRAGConfig` in lexical-graph.

### Core: `ByoKGConfig` singleton (`config.py`)
- `_ByoKGConfig` dataclass with lazy-evaluated properties and env var support
- `ResilientClient` wrapper for automatic credential refresh (SSO token expiry, expired sessions)
- Factory methods: `to_generator()`, `to_embedding()`, `to_reranker()`
- Single source of truth for model IDs, region, and generation settings

### Component auto-configuration
- `KGLinker` / `CypherKGLinker` — `llm_generator` optional, auto-pulls from config
- `AgenticRetriever` — `llm_generator` optional, auto-pulls from config
- `DenseIndex` / `NeptuneAnalyticsGraphStoreIndex` — `embedding` optional, auto-pulls from config
- `BedrockGenerator` — defaults from config when params are None; accepts `client=` for ResilientClient
- `LocalGReranker` — default model from config; improved error message for unsupported models

### Environment variables

| Variable | Description | Default |
|----------|-------------|---------|
| `BYOKG_LLM_MODEL` | LLM model ID | `anthropic.claude-sonnet-4-6` |
| `BYOKG_REGION` | AWS region override | boto3 default |
| `BYOKG_EMBED_MODEL` | Embedding model ID | `cohere.embed-english-v3` |
| `BYOKG_EMBED_DIMENSIONS` | Embedding dimensions | `1024` |
| `BYOKG_RERANKING_MODEL` | Reranking model ID | `BAAI/bge-reranker-base` |
| `BYOKG_MAX_TOKENS` | Max generation tokens | `4096` |
| `BYOKG_MAX_RETRIES` | Max retry attempts | `10` |

## Backward Compatibility

- All constructor signatures preserve original positional argument order
- Explicit parameter values override config defaults (no behavior change for existing code)
- No removed parameters, no renamed parameters
- No new required dependencies (uses stdlib + boto3/botocore already in requirements)

## Testing

- 31 new unit tests for config (singleton, defaults, env vars, setters, factory methods, ResilientClient)
- 344 total tests passing, 0 regressions
- Notebook integration test: 30/31 cells pass (1 GPU skip) on `byokg_rag_demo_local_graph.ipynb`

## Usage

```python
from graphrag_toolkit.byokg_rag import ByoKGConfig

# Configure once
ByoKGConfig.llm_model = "us.anthropic.claude-sonnet-4-6"
ByoKGConfig.region_name = "us-west-2"

# Factory methods
generator = ByoKGConfig.to_generator()
embedding = ByoKGConfig.to_embedding()

# Or let components auto-configure
linker = KGLinker(graph_store=my_store)  # generator created from config
index = DenseIndex()  # embedding created from config

# Or use environment variables (zero code changes)
# export BYOKG_LLM_MODEL=anthropic.claude-haiku-3
```

## Documentation

- Added "Global Configuration (ByoKGConfig)" section to `docs-site/src/content/docs/byokg-rag/configuration.mdx`

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: byokg-rag/src/graphrag_toolkit/byokg_rag/graph_connectors/kg_linker.py [hunk 1]
```diff
@@ -25,6 +25,8 @@ def __init__(self,
             graph_store: Component that provides access to graph data
             max_input_tokens: Maximum allowed tokens for inputs (default: 32000)
         """
+        if graph_store is None:
+            raise ValueError("graph_store is required. Pass a graph store instance.")
```
Comment (acarbonetto) [inline byokg-rag/src/graphrag_toolkit/byokg_rag/graph_connectors/kg_linker.py:29]: why...?  

What's the advantage of defaulting the parameter and then throwing a ValueError regardless?

File: byokg-rag/src/graphrag_toolkit/byokg_rag/config.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+    _llm_model: Optional[str] = None
+    _region_name: Optional[str] = field(default=None, init=False, repr=False)
+    # _region_checked distinguishes 'not yet read' from 'read and found None'
+    _region_checked: bool = field(default=False, init=False, repr=False)
+    _embed_model: Optional[str] = None
+    _embed_dimensions: Optional[int] = None
+    _reranking_model: Optional[str] = None
+    _max_tokens: Optional[int] = None
+    _max_retries: Optional[int] = None
+    _aws_clients: Dict = field(default_factory=dict)
```
Comment (acarbonetto) [inline byokg-rag/src/graphrag_toolkit/byokg_rag/config.py]:
```suggestion
    _aws_clients: Dict[str, 
  ResilientClient] = field(default_factory=dict)
```

File: byokg-rag/src/graphrag_toolkit/byokg_rag/config.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+                if isinstance(e, SSOTokenLoadError) or self._is_expired(e):
+                    logger.warning(f"[ResilientClient] Refreshing expired client for {self.service_name}")
+                    self._refresh_client()
+                    return getattr(self._client, name)(*args, **kwargs)
+                raise
+        return wrapper
+
+
+@dataclass
+class _ByoKGConfig:
```
Comment (acarbonetto) [inline byokg-rag/src/graphrag_toolkit/byokg_rag/config.py:90]: nit: consider adding a `reset()` method for testing

File: byokg-rag/tests/conftest.py [hunk 4]
```diff
@@ -7,6 +7,13 @@
 graph stores, LLM clients, and test data structures.
 """
 
+# Import faiss early to ensure its BLAS/OpenMP libraries load before torch's,
```
Comment (acarbonetto) [inline byokg-rag/tests/conftest.py]: this seems unrelated - maybe pull out into a separate PR?

File: byokg-rag/src/graphrag_toolkit/byokg_rag/graph_connectors/kg_linker.py [hunk 5]
```diff
@@ -135,8 +140,8 @@ class CypherKGLinker(KGLinker):
     This class focuses on generating and parsing LLM responses for cypher queries.
     """
     def __init__(self,
-            llm_generator, 
-            graph_store,
+            llm_generator=None, 
+            graph_store=None,
```
Comment (acarbonetto) [inline byokg-rag/src/graphrag_toolkit/byokg_rag/graph_connectors/kg_linker.py:144]: not sure if this is needed

File: byokg-rag/src/graphrag_toolkit/byokg_rag/graph_retrievers/graph_retrievers.py [hunk 6]
```diff
@@ -33,7 +33,7 @@ class AgenticRetriever(GRetriever):
     5. Continue until max iterations or early finish condition
     """
 
-    def __init__(self, llm_generator, graph_traversal, graph_verbalizer, pruning_reranker=None, 
+    def __init__(self, llm_generator=None, graph_traversal=None, graph_verbalizer=None, pruning_reranker=None,
```
Comment (acarbonetto) [inline byokg-rag/src/graphrag_toolkit/byokg_rag/graph_retrievers/graph_retrievers.py]: I'm not sure I understand the benefit of "graph_traversal=None, graph_verbalizer=None"

File: byokg-rag/src/graphrag_toolkit/byokg_rag/llm/bedrock_llms.py [hunk 7]
```diff
@@ -25,7 +25,7 @@ class BedrockGenerator(BaseGenerator):
     This class provides a wrapper around AWS Bedrock's Converse API for
     generating text responses using various foundation models.
     """
-    def __init__(self, model_name="global.anthropic.claude-sonnet-4-6", region_name="us-east-1", prefill=False, max_tokens = 4096, max_retries = 10, inference_config=None, reasoning_config=None):
+    def __init__(self, model_name=None, region_name=None, prefill=False, max_tokens=None, max_retries=None, inference_config=None, reasoning_config=None, boto3_session=None, client=None):
```
Comment (acarbonetto) [inline byokg-rag/src/graphrag_toolkit/byokg_rag/llm/bedrock_llms.py:28]: note: this will break for users that don't have their region set.  This is probably fine, but we should call it out in the description.

File: byokg-rag/src/graphrag_toolkit/byokg_rag/graph_connectors/kg_linker.py [hunk 8]
```diff
@@ -13,8 +13,8 @@ class KGLinker:
     """
 
     def __init__(self,
-            llm_generator, 
-            graph_store,
+            llm_generator=None,
```
Comment (acarbonetto) [inline byokg-rag/src/graphrag_toolkit/byokg_rag/graph_connectors/kg_linker.py:16]: Needs test for defaulted llm_generator

File: byokg-rag/src/graphrag_toolkit/byokg_rag/graph_connectors/kg_linker.py [hunk 9]
```diff
@@ -13,8 +13,8 @@ class KGLinker:
     """
 
     def __init__(self,
-            llm_generator, 
-            graph_store,
+            llm_generator=None, 
+            graph_store=None,
```
Comment (acarbonetto) [inline byokg-rag/src/graphrag_toolkit/byokg_rag/graph_connectors/kg_linker.py:17]: no test for defaulted graph_store

File: byokg-rag/src/graphrag_toolkit/byokg_rag/graph_retrievers/graph_retrievers.py (hunk 6 above)
Comment (acarbonetto) [inline byokg-rag/src/graphrag_toolkit/byokg_rag/graph_retrievers/graph_retrievers.py]: needs test for defaulted values

File: byokg-rag/src/graphrag_toolkit/byokg_rag/llm/bedrock_llms.py (hunk 7 above)
Comment (acarbonetto) [inline byokg-rag/src/graphrag_toolkit/byokg_rag/llm/bedrock_llms.py:28]: need test for defaulted values

File: byokg-rag/src/graphrag_toolkit/byokg_rag/llm/bedrock_llms.py [hunk 10]
```diff
@@ -62,13 +80,19 @@ def generate(self, prompt, system_prompt = "You are a helpful AI assistant.",  f
...
         """
-        response = generate_llm_response(self.region_name, self.model_name, system_prompt, prompt, self.max_new_tokens, self.max_retries, self.inference_config, self.reasoning_config)
+        response = generate_llm_response(self.region_name, self.model_name, system_prompt, prompt, self.max_new_tokens, self.max_retries, self.inference_config, self.reasoning_config, client=self.client)
         if "Failed due to other reasons." in response:
             raise Exception(f"{response}")
         return response
         
-def generate_llm_response(region_name, model_id, system_prompt, query, max_tokens, max_retries, inference_config=None, reasoning_config=None):
-    bedrock_runtime = boto3.client("bedrock-runtime", region_name=region_name)
+def generate_llm_response(region_name, model_id, system_prompt, query, max_tokens, max_retries, inference_config=None, reasoning_config=None, *, client=None):
```
Comment (acarbonetto) [inline byokg-rag/src/graphrag_toolkit/byokg_rag/llm/bedrock_llms.py:90]: missing test for client=None

Comment (acarbonetto) [reply byokg-rag/src/graphrag_toolkit/byokg_rag/llm/bedrock_llms.py:90]: And test for region_name=none

File: byokg-rag/src/graphrag_toolkit/byokg_rag/graph_connectors/kg_linker.py (hunk 1 above)
Comment (andy-k-improving) [reply byokg-rag/src/graphrag_toolkit/byokg_rag/graph_connectors/kg_linker.py:29]: | Caller does | `def foo(graph_store=None)` + check | `def foo(graph_store)` + check |
|---|---|---|
| `foo()` (omits arg) | ✅ Hits check → clear `ValueError` | ❌ `TypeError: missing argument` (Python catches it) |
| `foo(None)` | ✅ Hits check → clear `ValueError` | ✅ Hits check → clear `ValueError` |
| `foo(valid_store)` | ✅ Works | ✅ Works |


I guess this is aiming to providing a more detail error message in the ❌ case on above table, as Python won't safe guard `None` entry on runtime regardless.

Comment (mykola-pereyma) [reply byokg-rag/src/graphrag_toolkit/byokg_rag/graph_connectors/kg_linker.py:29]: Yeap, that is the goal.

File: byokg-rag/tests/conftest.py (hunk 4 above)
Comment (mykola-pereyma) [reply byokg-rag/tests/conftest.py]: extracted to #273

File: byokg-rag/src/graphrag_toolkit/byokg_rag/graph_connectors/kg_linker.py (hunk 5 above)
Comment (mykola-pereyma) [reply byokg-rag/src/graphrag_toolkit/byokg_rag/graph_connectors/kg_linker.py:144]: Making `llm_generator` optional is the core intent of this PR — it enables components to auto-resolve their LLM from the centralized `ByoKGConfig` singleton instead of requiring manual wiring at every call site. Users can still pass an explicit generator to override the default.

File: byokg-rag/src/graphrag_toolkit/byokg_rag/llm/bedrock_llms.py (hunk 7 above)
Comment (mykola-pereyma) [reply byokg-rag/src/graphrag_toolkit/byokg_rag/llm/bedrock_llms.py:28]: added docstring describing options.

File: byokg-rag/src/graphrag_toolkit/byokg_rag/graph_retrievers/graph_retrievers.py (hunk 6 above)
Comment (mykola-pereyma) [reply byokg-rag/src/graphrag_toolkit/byokg_rag/graph_retrievers/graph_retrievers.py]: agree, will make them positional

**Review by acarbonetto — APPROVED** [review]
